apps/ledger-live-mobile/src/components/RequiresBLE/hooks/useAndroidBluetoothPermissions.ts

```diff
@@ -0,0 +1,191 @@
+import {
+  Permission,
+  PermissionsAndroid,
+  Platform,
+  AppState,
+} from "react-native";
+import { useRef, useCallback, useState, useEffect } from "react";
+
+const { PERMISSIONS, RESULTS } = PermissionsAndroid;
+
+// No check is done on iOS here
+export const bluetoothPermissions: Permission[] =
+  Platform.OS === "android"
+    ? Platform.Version < 31
+      ? []
+      : [PERMISSIONS.BLUETOOTH_SCAN, PERMISSIONS.BLUETOOTH_CONNECT]
+    : [];
+
+/**
+ * Checks if the app has the required permissions to use Bluetooth.
+ *
+ * On Android 11 and above, the app needs to have the BLUETOOTH_SCAN and BLUETOOTH_CONNECT permissions.
+ * On iOS, no check is done and returns true.
+ */
+async function checkBluetoothPermissions() {
+  if (bluetoothPermissions.length === 0) return true;
+
+  return Promise.all(
+    bluetoothPermissions.map(permission =>
+      PermissionsAndroid.check(permission),
+    ),
+  ).then(results => results.every(res => res));
+}
+
+type RequestMultipleResult = {
+  allGranted: boolean;
+  generalStatus: typeof PermissionsAndroid.RESULTS[number];
+};
+
+/**
+ * Requests the required permissions to use Bluetooth.
+ *
+ * On Android 11 and above, the app needs to have the BLUETOOTH_SCAN and BLUETOOTH_CONNECT permissions.
+ * On iOS, no request is done and returns true.
+ */
+async function requestBluetoothPermissions(): Promise<RequestMultipleResult> {
+  if (bluetoothPermissions.length === 0)
+    return { allGranted: true, generalStatus: RESULTS.GRANTED };
+
+  return PermissionsAndroid.requestMultiple(bluetoothPermissions).then(res => {
+    const permissionStatuses = Object.values(res);
+    const allGranted = permissionStatuses.every(
+      status => status === RESULTS.GRANTED,
+    );
+
+    /** https://developer.android.com/about/versions/11/privacy/permissions#dialog-visibility */
+    const someNeverAskAgain = permissionStatuses.some(
+      status => status === RESULTS.NEVER_ASK_AGAIN,
+    );
+
+    return {
+      allGranted,
+      generalStatus: allGranted
+        ? RESULTS.GRANTED
+        : someNeverAskAgain
+        ? RESULTS.NEVER_ASK_AGAIN
+        : RESULTS.DENIED,
+    };
+  });
+}
+
+/**
+ * Hook to check and request the BLE permission on Android.
+ *
+ * @returns an object with the following properties:
+ * - hasPermission: true if the permission is granted, false if it's denied, undefined if it's still loading
+ * - checkAgain: a function to check (and not request) again the permission
+ * - neverAskAgain: true if the user has checked the "never ask again" checkbox (or on Android 11+ if the user has denied the permission several times)
+ * - renderChildren: a boolean to know if the children should be rendered (permissions are granted) or not
+ */
+export const useAndroidBluetoothPermissions = () => {
+  const [hasPermission, setHasPermission] = useState<boolean | undefined>();
+  const [check, setCheck] = useState(false);
```

Comment:
I think the naming of "hasPermission" and "check" is not clear (check being a verb, while it's a boolean, maybe checkResult is more appropriate?), and maybe the structure (separating in two states) is weird.
Why are we separating the result of the permission check and the result of the permission request? In the end these two resolve in a state that means "the permissions is / isn't granted", no?
Also i see that `check` and `hasPermission`  are only used in one place, like this: `check || hasPermission` so I feel like we can get rid of one of them.

If you what you want to track is whether the initial check has resolved, then track it in a state with a name that clearly illustrates that.

Reply:
Yeah i agree, i'm going to re-work this hook 👌